File: backend/src/onboarding/onboarding.service.ts
import { BadRequestException, Injectable, NotFoundException } from "@nestjs/common";
import { InjectRepository } from "@nestjs/typeorm";
import { DataSource, EntityManager, Repository } from "typeorm";
import * as bcrypt from "bcrypt";
import { randomUUID } from "crypto";
import { AuditService } from "../audit/audit.service";
import { Tenant, TenantStatus } from "../tenants/tenant.entity";
import { School } from "../schools/school.entity";
import { Campus } from "../campuses/campus.entity";
import {
  AcademicYear,
  ClassArm,
  ClassLevel,
  GradingScheme,
  Subject,
  Term,
} from "../academic/academic.entity";
import { User } from "../users/user.entity";
import { Device } from "../devices/device.entity";
import { BootstrapSchoolSetupDto } from "./dto/bootstrap-school.dto";

@Injectable()
export class OnboardingService {
  constructor(
    private readonly dataSource: DataSource,
    @InjectRepository(User) private readonly users: Repository<User>,
    private readonly audit: AuditService,
  ) {}

  async bootstrapSchoolSetup(
    currentUserId: string,
    dto: BootstrapSchoolSetupDto,
  ) {
    const currentUser = await this.users.findOne({
      where: { id: currentUserId, deleted: false, isActive: true },
    });

    if (!currentUser) {
      throw new BadRequestException("Authenticated user not found.");
    }

    if (currentUser.tenantId || currentUser.schoolId || currentUser.campusId) {
      if (!currentUser.tenantId || !currentUser.schoolId || !currentUser.campusId) {
        throw new BadRequestException(
          "School setup is in an invalid partial state for this user.",
        );
      }
      return this.resumeBootstrapSchoolSetup(currentUser, dto);
    }

    const userId = currentUser.id;
    const shouldRegisterTrustedDevice =
      dto.deviceRegistration.registerOfflineAccess;
    if (
      shouldRegisterTrustedDevice &&
      (!dto.deviceRegistration.deviceFingerprint ||
        !dto.deviceRegistration.deviceName)
    ) {
      throw new BadRequestException(
        "Trusted device registration requires a device name and fingerprint.",
      );
    }

    return this.dataSource.transaction(async (manager) => {
      const tenant = await manager.save(
        Tenant,
        manager.create(Tenant, {
          name: dto.school.name,
          status: TenantStatus.Trial,
          contactEmail: dto.school.contactEmail ?? null,
          contactPhone: dto.school.contactPhone ?? null,
          deleted: false,
        }),
      );

      const school = await manager.save(
        School,
        manager.create(School, {
          tenantId: tenant.id,
          name: dto.school.name,
          shortName: dto.school.shortName ?? null,
          schoolType: dto.school.schoolType,
          address: dto.school.address ?? null,
          region: dto.school.region ?? null,
          district: dto.school.district ?? null,
          contactPhone: dto.school.contactPhone ?? null,
          contactEmail: dto.school.contactEmail ?? null,
          onboardingDefaults: {
            staffRoles: dto.onboardingDefaults.staffRoles,
            feeCategories: dto.onboardingDefaults.feeCategories,
            receiptFormat: dto.onboardingDefaults.receiptFormat,
            notifications: dto.onboardingDefaults.notifications,
          },
          serverRevision: await this.nextServerRevision(manager),
          deleted: false,
        }),
      );

      const campus = await manager.save(
        Campus,
        manager.create(Campus, {
          tenantId: tenant.id,
          schoolId: school.id,
          name: dto.campus.name,
          address: dto.campus.address ?? null,
          contactPhone: dto.campus.contactPhone ?? null,
          registrationCode: dto.campus.registrationCode ?? null,
          serverRevision: await this.nextServerRevision(manager),
          deleted: false,
        }),
      );

      const year = await manager.save(
        AcademicYear,
        manager.create(AcademicYear, {
          tenantId: tenant.id,
          schoolId: school.id,
          label: dto.academicYear.label,
          startDate: dto.academicYear.startDate,
          endDate: dto.academicYear.endDate,
          isCurrent: true,
          serverRevision: await this.nextServerRevision(manager),
          deleted: false,
        }),
      );

      const terms: Term[] = [];
      for (const term of dto.academicYear.terms) {
        const savedTerm = await manager.save(
          Term,
          manager.create(Term, {
            tenantId: tenant.id,
            schoolId: school.id,
            academicYearId: year.id,
            name: term.name,
            termNumber: term.termNumber,
            startDate: term.startDate,
            endDate: term.endDate,
            isCurrent: term.isCurrent,
            serverRevision: await this.nextServerRevision(manager),
            deleted: false,
          }),
        );
        terms.push(savedTerm);
      }

      const classLevels: ClassLevel[] = [];
      const classArms: ClassArm[] = [];
      for (const levelDto of dto.classLevels) {
        const level = await manager.save(
          ClassLevel,
          manager.create(ClassLevel, {
            tenantId: tenant.id,
            schoolId: school.id,
            name: levelDto.name,
            sortOrder: levelDto.sortOrder,
            serverRevision: await this.nextServerRevision(manager),
            deleted: false,
          }),
        );
        classLevels.push(level);

        for (const armDto of levelDto.arms) {
          const classArm = await manager.save(
            ClassArm,
            manager.create(ClassArm, {
              tenantId: tenant.id,
              schoolId: school.id,
              classLevelId: level.id,
              arm: armDto.arm,
              displayName: `${levelDto.name} ${armDto.arm}`,
              serverRevision: await this.nextServerRevision(manager),
              deleted: false,
            }),
          );
          classArms.push(classArm);
        }
      }

      const subjects: Subject[] = [];
      for (const subjectDto of dto.subjects) {
        const subject = await manager.save(
          Subject,
          manager.create(Subject, {
            tenantId: tenant.id,
            schoolId: school.id,
            name: subjectDto.name,
            code: subjectDto.code ?? null,
            serverRevision: await this.nextServerRevision(manager),
            deleted: false,
          }),
        );
        subjects.push(subject);
      }

      const gradingScheme = await manager.save(
        GradingScheme,
        manager.create(GradingScheme, {
          tenantId: tenant.id,
          schoolId: school.id,
          name: dto.gradingScheme.name,
          bands: dto.gradingScheme.bands,
          isDefault: true,
          serverRevision: await this.nextServerRevision(manager),
          deleted: false,
        }),
      );

      await manager.update(User, userId, {
        tenantId: tenant.id,
        schoolId: school.id,
        campusId: campus.id,
      });

      const deviceRegistration = shouldRegisterTrustedDevice
        ? await this.registerTrustedDevice(manager, {
            tenantId: tenant.id,
            schoolId: school.id,
            campusId: campus.id,
            userId,
            deviceName: dto.deviceRegistration.deviceName,
            deviceFingerprint: dto.deviceRegistration.deviceFingerprint,
          })
        : null;

      const updatedUser = await manager.findOne(User, {
        where: { id: userId },
      });

      return this.buildBootstrapResponse({
        tenant,
        school,
        campus,
        academicYear: year,
        terms,
        classLevels,
        classArms,
        subjects,
        gradingScheme,
        user: updatedUser,
        deviceRegistration,
      });
    });
  }

  private async resumeBootstrapSchoolSetup(
    currentUser: User,
    dto: BootstrapSchoolSetupDto,
  ) {
    return this.dataSource.transaction(async (manager) => {
      const tenant = await manager.findOne(Tenant, {
        where: { id: currentUser.tenantId!, deleted: false },
      });
      const school = await manager.findOne(School, {
        where: {
          id: currentUser.schoolId!,
          tenantId: currentUser.tenantId!,
          deleted: false,
        },
      });
      const campus = await manager.findOne(Campus, {
        where: {
          id: currentUser.campusId!,
          tenantId: currentUser.tenantId!,
          schoolId: currentUser.schoolId!,
          deleted: false,
        },
      });

      if (!tenant || !school || !campus) {
        throw new NotFoundException(
          "Existing onboarding workspace could not be reconstructed.",
        );
      }

      const academicYear = await manager.findOne(AcademicYear, {
        where: {
          tenantId: currentUser.tenantId!,
          schoolId: currentUser.schoolId!,
          deleted: false,
          isCurrent: true,
        },
        order: {
          updatedAt: "DESC",
        },
      });
      const terms = academicYear
        ? await manager.find(Term, {
            where: {
              tenantId: currentUser.tenantId!,
              schoolId: currentUser.schoolId!,
              academicYearId: academicYear.id,
              deleted: false,
            },
            order: {
              termNumber: "ASC",
            },
          })
        : [];
      const classLevels = await manager.find(ClassLevel, {
        where: {
          tenantId: currentUser.tenantId!,
          schoolId: currentUser.schoolId!,
          deleted: false,
        },
        order: {
          sortOrder: "ASC",
          createdAt: "ASC",
        },
      });
      const classArms = await manager.find(ClassArm, {
        where: {
          tenantId: currentUser.tenantId!,
          schoolId: currentUser.schoolId!,
          deleted: false,
        },
        order: {
          displayName: "ASC",
        },
      });
      const subjects = await manager.find(Subject, {
        where: {
          tenantId: currentUser.tenantId!,
          schoolId: currentUser.schoolId!,
          deleted: false,
        },
        order: {
          name: "ASC",
        },
      });
      const gradingScheme =
        (await manager.findOne(GradingScheme, {
          where: {
            tenantId: currentUser.tenantId!,
            schoolId: currentUser.schoolId!,
            deleted: false,
            isDefault: true,
          },
          order: {
            updatedAt: "DESC",
          },
        })) ??
        (await manager.findOne(GradingScheme, {
          where: {
            tenantId: currentUser.tenantId!,
            schoolId: currentUser.schoolId!,
            deleted: false,
          },
          order: {
            updatedAt: "DESC",
          },
        }));

      const deviceRegistration = dto.deviceRegistration.registerOfflineAccess
        ? await this.registerTrustedDevice(manager, {
            tenantId: tenant.id,
            schoolId: school.id,
            campusId: campus.id,
            userId: currentUser.id,
            deviceName: dto.deviceRegistration.deviceName,
            deviceFingerprint: dto.deviceRegistration.deviceFingerprint,
          })
        : null;

      return this.buildBootstrapResponse({
        tenant,
        school,
        campus,
        academicYear,
        terms,
        classLevels,
        classArms,
        subjects,
        gradingScheme,
        user: currentUser,
        deviceRegistration,
      });
    });
  }

  private async registerTrustedDevice(
    manager: EntityManager,
    input: {
      tenantId: string;
      schoolId: string;
      campusId: string;
      userId: string;
      deviceName: string;
      deviceFingerprint: string;
    },
  ) {
    const rawOfflineToken = `${randomUUID()}-${randomUUID()}`;
    const offlineTokenHash = await bcrypt.hash(rawOfflineToken, 12);
    const existingDevice = await manager.findOne(Device, {
      where: {
        deviceFingerprint: input.deviceFingerprint,
      },
    });

    if (
      existingDevice &&
      existingDevice.registeredByUserId !== input.userId &&
      existingDevice.isActive
    ) {
      throw new BadRequestException(
        "Device fingerprint is already registered to another user.",
      );
    }

    const device = await manager.save(
      Device,
      manager.create(Device, {
        ...(existingDevice ?? {}),
        id: existingDevice?.id,
        deviceName: input.deviceName,
        deviceFingerprint: input.deviceFingerprint,
        offlineTokenHash,
        tenantId: input.tenantId,
        schoolId: input.schoolId,
        campusId: input.campusId,
        registeredByUserId: input.userId,
        isActive: true,
      }),
    );

    await this.audit.record(
      {
        tenantId: input.tenantId,
        schoolId: input.schoolId,
        campusId: input.campusId,
        actorUserId: input.userId,
        eventType: "devices.trusted_device_registered",
        entityType: "device",
        entityId: device.id,
        metadata: {
          deviceName: device.deviceName,
          deviceFingerprint: device.deviceFingerprint,
          mode:
            existingDevice == null
              ? "onboarding_registration"
              : existingDevice.registeredByUserId === input.userId
                ? "onboarding_credential_rotation"
                : "onboarding_reassigned_after_revoke",
        },
      },
      manager,
    );

    return {
      deviceId: device.id,
      deviceName: device.deviceName,
      deviceFingerprint: device.deviceFingerprint,
      offlineToken: rawOfflineToken,
    };
  }

  private buildBootstrapResponse(input: {
    tenant: Tenant;
    school: School;
    campus: Campus;
    academicYear: AcademicYear | null;
    terms: Term[];
    classLevels: ClassLevel[];
    classArms: ClassArm[];
    subjects: Subject[];
    gradingScheme: GradingScheme | null;
    user: User | null;
    deviceRegistration:
      | {
          deviceId: string;
          deviceName: string;
          deviceFingerprint: string;
          offlineToken: string;
        }
      | null;
  }) {
    return {
      tenant: {
        id: input.tenant.id,
        name: input.tenant.name,
        status: input.tenant.status,
        contactEmail: input.tenant.contactEmail,
        contactPhone: input.tenant.contactPhone,
        deleted: input.tenant.deleted,
        createdAt: input.tenant.createdAt.toISOString(),
        updatedAt: input.tenant.updatedAt.toISOString(),
      },
      school: {
        id: input.school.id,
        tenantId: input.school.tenantId,
        name: input.school.name,
        shortName: input.school.shortName,
        schoolType: input.school.schoolType,
        address: input.school.address,
        region: input.school.region,
        district: input.school.district,
        contactPhone: input.school.contactPhone,
        contactEmail: input.school.contactEmail,
        onboardingDefaults: input.school.onboardingDefaults,
        serverRevision: input.school.serverRevision,
        deleted: input.school.deleted,
        createdAt: input.school.createdAt.toISOString(),
        updatedAt: input.school.updatedAt.toISOString(),
      },
      campus: {
        id: input.campus.id,
        tenantId: input.campus.tenantId,
        schoolId: input.campus.schoolId,
        name: input.campus.name,
        address: input.campus.address,
        contactPhone: input.campus.contactPhone,
        registrationCode: input.campus.registrationCode,
        serverRevision: input.campus.serverRevision,
        deleted: input.campus.deleted,
        createdAt: input.campus.createdAt.toISOString(),
        updatedAt: input.campus.updatedAt.toISOString(),
      },
      bootstrapSnapshot: {
        academicYear: input.academicYear
          ? this.serializeEntity(input.academicYear)
          : null,
        terms: input.terms.map((term) => this.serializeEntity(term)),
        classLevels: input.classLevels.map((level) => this.serializeEntity(level)),
        classArms: input.classArms.map((arm) => this.serializeEntity(arm)),
        subjects: input.subjects.map((subject) => this.serializeEntity(subject)),
        gradingScheme: input.gradingScheme
          ? this.serializeEntity(input.gradingScheme)
          : null,
      },
      user: input.user
        ? {
            id: input.user.id,
            email: input.user.email,
            fullName: input.user.fullName,
            role: input.user.role,
            tenantId: input.user.tenantId,
            schoolId: input.user.schoolId,
            campusId: input.user.campusId,
          }
        : null,
      deviceRegistration: input.deviceRegistration,
    };
  }

  private serializeEntity<T extends { createdAt: Date; updatedAt: Date }>(
    entity: T,
  ) {
    return {
      ...entity,
      createdAt: entity.createdAt.toISOString(),
      updatedAt: entity.updatedAt.toISOString(),
    };
  }

  private async nextServerRevision(manager: {
    query: (query: string) => Promise<unknown>;
  }) {
    const result = (await manager.query(
      "SELECT nextval('sync_server_revision_seq')::bigint AS revision",
    )) as { revision: string | number }[];
    return Number(result[0].revision);
  }
}

File: apps/desktop_app/lib/ui/onboarding/onboarding_service.dart
import 'dart:convert';

import 'package:dio/dio.dart';
import 'package:drift/drift.dart';
import 'package:uuid/uuid.dart';

import 'package:desktop_app/auth/auth_service.dart';
import 'package:desktop_app/database/app_database.dart';
import 'package:desktop_app/ui/onboarding/onboarding_models.dart';

class OnboardingService {
  OnboardingService(this._auth, this._db);

  final AuthService _auth;
  final AppDatabase _db;
  final _uuid = const Uuid();

  Future<AuthUser> bootstrapSchool(OnboardingDraft draft) async {
    final deviceFingerprint = draft.deviceRegistration.registerOfflineAccess
        ? await _auth.ensureDeviceFingerprint()
        : null;
    final dio = _auth.createAuthenticatedClient();
    final response = await dio.post<Map<String, dynamic>>(
      '/onboarding/bootstrap-school',
      data: {
        'school': {
          'name': draft.school.name,
          'shortName': _nullIfBlank(draft.school.shortName),
          'schoolType': draft.school.schoolType,
          'address': _nullIfBlank(draft.school.address),
          'region': _nullIfBlank(draft.school.region),
          'district': _nullIfBlank(draft.school.district),
          'contactPhone': _nullIfBlank(draft.school.contactPhone),
          'contactEmail': _nullIfBlank(draft.school.contactEmail),
        },
        'campus': {
          'name': draft.campus.name,
          'address': _nullIfBlank(draft.campus.address),
          'contactPhone': _nullIfBlank(draft.campus.contactPhone),
          'registrationCode': _nullIfBlank(draft.campus.registrationCode),
        },
        'academicYear': {
          'label': draft.academicYear.label,
          'startDate': draft.academicYear.startDate,
          'endDate': draft.academicYear.endDate,
          'terms': draft.academicYear.terms
              .map(
                (term) => {
                  'name': term.name,
                  'termNumber': term.termNumber,
                  'startDate': term.startDate,
                  'endDate': term.endDate,
                  'isCurrent': term.isCurrent,
                },
              )
              .toList(),
        },
        'classLevels': draft.classSetup.levels
            .map(
              (level) => {
                'name': level.name,
                'sortOrder': level.sortOrder,
                'arms': level.arms.map((arm) => {'arm': arm}).toList(),
              },
            )
            .toList(),
        'subjects': draft.classSetup.subjects
            .map(
              (subject) => {
                'name': subject.name,
                'code': _nullIfBlank(subject.code),
              },
            )
            .toList(),
        'gradingScheme': {
          'name': draft.gradingScheme.name,
          'bands': draft.gradingScheme.bands
              .map(
                (band) => {
                  'grade': band.grade,
                  'min': band.min,
                  'max': band.max,
                  'remark': band.remark,
                },
              )
              .toList(),
        },
        'onboardingDefaults': {
          'staffRoles': draft.staffRoles
              .map(
                (role) => {
                  'role': role.role,
                  'enabled': role.enabled,
                  'headcount': role.headcount,
                },
              )
              .toList(),
          'feeCategories': draft.feeCategories
              .map(
                (category) => {
                  'name': category.name,
                  'defaultAmount': category.defaultAmount,
                  'billingTerm': category.billingTerm,
                },
              )
              .toList(),
          'receiptFormat': {
            'headerLine1': _nullIfBlank(draft.receiptFormat.headerLine1),
            'headerLine2': _nullIfBlank(draft.receiptFormat.headerLine2),
            'footerNote': _nullIfBlank(draft.receiptFormat.footerNote),
            'receiptPrefix': draft.receiptFormat.receiptPrefix,
            'nextReceiptNumber': draft.receiptFormat.nextReceiptNumber,
          },
          'notifications': {
            'smsEnabled': draft.notifications.smsEnabled,
            'paymentReceiptsEnabled':
                draft.notifications.paymentReceiptsEnabled,
            'feeRemindersEnabled': draft.notifications.feeRemindersEnabled,
            'senderId': _nullIfBlank(draft.notifications.senderId),
            'providerName': _nullIfBlank(draft.notifications.providerName),
          },
        },
        'deviceRegistration': {
          'registerOfflineAccess':
              draft.deviceRegistration.registerOfflineAccess,
          'deviceName': _nullIfBlank(
                draft.deviceRegistration.deviceName,
              ) ??
              'Offline School Desktop',
          'deviceFingerprint': deviceFingerprint,
        },
      },
    );

    final userJson = response.data?['user'] as Map<String, dynamic>?;
    if (userJson == null) {
      throw DioException(
        requestOptions: response.requestOptions,
        error: 'Bootstrap response did not include user context.',
      );
    }

    final user = _auth.updateCurrentUserFromJson(userJson);
    await _syncTrustedDeviceState(response.data ?? const {}, user);
    await _seedBootstrapSnapshot(response.data ?? const {}, user);
    await _seedOnboardingFinanceDefaults(draft, user);
    return user;
  }

  String? _nullIfBlank(String value) {
    final trimmed = value.trim();
    return trimmed.isEmpty ? null : trimmed;
  }

  Future<void> _seedBootstrapSnapshot(
    Map<String, dynamic> responseData,
    AuthUser user,
  ) async {
    final tenantId = user.tenantId;
    final schoolId = user.schoolId;
    if (tenantId == null || schoolId == null) {
      return;
    }

    final scope = LocalDataScope(
      tenantId: tenantId,
      schoolId: schoolId,
      campusId: user.campusId,
    );
    await _db.reconcileLocalScope(scope: scope);

    final snapshot =
        responseData['bootstrapSnapshot'] as Map<String, dynamic>? ?? const {};
    final tenant = responseData['tenant'] as Map<String, dynamic>? ?? const {};
    if (tenant.isNotEmpty) {
      await _db.upsertTenantProfile(
        TenantProfileCacheCompanion(
          id: Value('${tenant['id'] ?? tenantId}'),
          name: Value('${tenant['name']}'),
          status: Value('${tenant['status'] ?? 'trial'}'),
          contactEmail: Value(tenant['contactEmail'] as String?),
          contactPhone: Value(tenant['contactPhone'] as String?),
          deleted: Value(tenant['deleted'] as bool? ?? false),
          createdAt: Value(DateTime.parse('${tenant['createdAt']}')),
          updatedAt: Value(DateTime.parse('${tenant['updatedAt']}')),
        ),
      );
    }

    final school = responseData['school'] as Map<String, dynamic>? ?? const {};
    if (school.isNotEmpty) {
      await _db.upsertSchoolProfile(
        SchoolProfileCacheCompanion(
          id: Value('${school['id']}'),
          tenantId: Value('${school['tenantId'] ?? tenantId}'),
          name: Value('${school['name']}'),
          shortName: Value(school['shortName'] as String?),
          schoolType: Value('${school['schoolType']}'),
          address: Value(school['address'] as String?),
          region: Value(school['region'] as String?),
          district: Value(school['district'] as String?),
          contactPhone: Value(school['contactPhone'] as String?),
          contactEmail: Value(school['contactEmail'] as String?),
          onboardingDefaultsJson: Value(
            jsonEncode(school['onboardingDefaults'] ?? const {}),
          ),
          serverRevision: Value(school['serverRevision'] as int? ?? 0),
          deleted: Value(school['deleted'] as bool? ?? false),
          createdAt: Value(DateTime.parse('${school['createdAt']}')),
          updatedAt: Value(DateTime.parse('${school['updatedAt']}')),
        ),
      );
      await _db.updateLastRevision(
        'school',
        school['serverRevision'] as int? ?? 0,
      );
    }

    final campus = responseData['campus'] as Map<String, dynamic>? ?? const {};
    if (campus.isNotEmpty) {
      await _db.upsertCampusProfile(
        CampusProfileCacheCompanion(
          id: Value('${campus['id']}'),
          tenantId: Value('${campus['tenantId'] ?? tenantId}'),
          schoolId: Value('${campus['schoolId'] ?? schoolId}'),
          name: Value('${campus['name']}'),
          address: Value(campus['address'] as String?),
          contactPhone: Value(campus['contactPhone'] as String?),
          registrationCode: Value(campus['registrationCode'] as String?),
          serverRevision: Value(campus['serverRevision'] as int? ?? 0),
          deleted: Value(campus['deleted'] as bool? ?? false),
          createdAt: Value(DateTime.parse('${campus['createdAt']}')),
          updatedAt: Value(DateTime.parse('${campus['updatedAt']}')),
        ),
      );
      await _db.updateLastRevision(
        'campus',
        campus['serverRevision'] as int? ?? 0,
      );
    }

    final academicYear =
        snapshot['academicYear'] as Map<String, dynamic>? ?? const {};
    if (academicYear.isNotEmpty) {
      await _db.upsertAcademicYear(
        AcademicYearsCacheCompanion(
          id: Value('${academicYear['id']}'),
          tenantId: Value('${academicYear['tenantId']}'),
          schoolId: Value('${academicYear['schoolId']}'),
          label: Value('${academicYear['label']}'),
          startDate: Value('${academicYear['startDate']}'),
          endDate: Value('${academicYear['endDate']}'),
          isCurrent: Value(academicYear['isCurrent'] as bool? ?? false),
          serverRevision: Value(academicYear['serverRevision'] as int? ?? 0),
          deleted: Value(academicYear['deleted'] as bool? ?? false),
          createdAt: Value(DateTime.parse('${academicYear['createdAt']}')),
          updatedAt: Value(DateTime.parse('${academicYear['updatedAt']}')),
        ),
      );
      await _db.updateLastRevision(
        'academic_year',
        academicYear['serverRevision'] as int? ?? 0,
      );
    }

    var maxTermRevision = 0;
    for (final item in snapshot['terms'] as List<dynamic>? ?? const []) {
      final term = Map<String, dynamic>.from(item as Map);
      final revision = term['serverRevision'] as int? ?? 0;
      await _db.upsertTerm(
        TermsCacheCompanion(
          id: Value('${term['id']}'),
          tenantId: Value('${term['tenantId']}'),
          schoolId: Value('${term['schoolId']}'),
          academicYearId: Value('${term['academicYearId']}'),
          name: Value('${term['name']}'),
          termNumber: Value(term['termNumber'] as int? ?? 0),
          startDate: Value('${term['startDate']}'),
          endDate: Value('${term['endDate']}'),
          isCurrent: Value(term['isCurrent'] as bool? ?? false),
          serverRevision: Value(revision),
          deleted: Value(term['deleted'] as bool? ?? false),
          createdAt: Value(DateTime.parse('${term['createdAt']}')),
          updatedAt: Value(DateTime.parse('${term['updatedAt']}')),
        ),
      );
      if (revision > maxTermRevision) {
        maxTermRevision = revision;
      }
    }
    if (maxTermRevision > 0) {
      await _db.updateLastRevision('term', maxTermRevision);
    }

    var maxClassLevelRevision = 0;
    for (final item in snapshot['classLevels'] as List<dynamic>? ?? const []) {
      final level = Map<String, dynamic>.from(item as Map);
      final revision = level['serverRevision'] as int? ?? 0;
      await _db.upsertClassLevel(
        ClassLevelsCacheCompanion(
          id: Value('${level['id']}'),
          tenantId: Value('${level['tenantId']}'),
          schoolId: Value('${level['schoolId']}'),
          name: Value('${level['name']}'),
          sortOrder: Value(level['sortOrder'] as int? ?? 0),
          serverRevision: Value(revision),
          deleted: Value(level['deleted'] as bool? ?? false),
          createdAt: Value(DateTime.parse('${level['createdAt']}')),
          updatedAt: Value(DateTime.parse('${level['updatedAt']}')),
        ),
      );
      if (revision > maxClassLevelRevision) {
        maxClassLevelRevision = revision;
      }
    }
    if (maxClassLevelRevision > 0) {
      await _db.updateLastRevision('class_level', maxClassLevelRevision);
    }

    var maxClassArmRevision = 0;
    for (final item in snapshot['classArms'] as List<dynamic>? ?? const []) {
      final arm = Map<String, dynamic>.from(item as Map);
      final revision = arm['serverRevision'] as int? ?? 0;
      await _db.upsertClassArm(
        ClassArmsCacheCompanion(
          id: Value('${arm['id']}'),
          tenantId: Value('${arm['tenantId']}'),
          schoolId: Value('${arm['schoolId']}'),
          classLevelId: Value('${arm['classLevelId']}'),
          arm: Value('${arm['arm']}'),
          displayName: Value('${arm['displayName']}'),
          serverRevision: Value(revision),
          deleted: Value(arm['deleted'] as bool? ?? false),
          createdAt: Value(DateTime.parse('${arm['createdAt']}')),
          updatedAt: Value(DateTime.parse('${arm['updatedAt']}')),
        ),
      );
      if (revision > maxClassArmRevision) {
        maxClassArmRevision = revision;
      }
    }
    if (maxClassArmRevision > 0) {
      await _db.updateLastRevision('class_arm', maxClassArmRevision);
    }

    var maxSubjectRevision = 0;
    for (final item in snapshot['subjects'] as List<dynamic>? ?? const []) {
      final subject = Map<String, dynamic>.from(item as Map);
      final revision = subject['serverRevision'] as int? ?? 0;
      await _db.upsertSubject(
        SubjectsCacheCompanion(
          id: Value('${subject['id']}'),
          tenantId: Value('${subject['tenantId']}'),
          schoolId: Value('${subject['schoolId']}'),
          name: Value('${subject['name']}'),
          code: Value(subject['code'] as String?),
          serverRevision: Value(revision),
          deleted: Value(subject['deleted'] as bool? ?? false),
          createdAt: Value(DateTime.parse('${subject['createdAt']}')),
          updatedAt: Value(DateTime.parse('${subject['updatedAt']}')),
        ),
      );
      if (revision > maxSubjectRevision) {
        maxSubjectRevision = revision;
      }
    }
    if (maxSubjectRevision > 0) {
      await _db.updateLastRevision('subject', maxSubjectRevision);
    }

    final gradingScheme =
        snapshot['gradingScheme'] as Map<String, dynamic>? ?? const {};
    if (gradingScheme.isNotEmpty) {
      final revision = gradingScheme['serverRevision'] as int? ?? 0;
      await _db.upsertGradingScheme(
        GradingSchemesCacheCompanion(
          id: Value('${gradingScheme['id']}'),
          tenantId: Value('${gradingScheme['tenantId']}'),
          schoolId: Value('${gradingScheme['schoolId']}'),
          name: Value('${gradingScheme['name']}'),
          bandsJson: Value(jsonEncode(gradingScheme['bands'])),
          isDefault: Value(gradingScheme['isDefault'] as bool? ?? false),
          serverRevision: Value(revision),
          deleted: Value(gradingScheme['deleted'] as bool? ?? false),
          createdAt: Value(DateTime.parse('${gradingScheme['createdAt']}')),
          updatedAt: Value(DateTime.parse('${gradingScheme['updatedAt']}')),
        ),
      );
      await _db.updateLastRevision('grading_scheme', revision);
    }
  }

  Future<void> _syncTrustedDeviceState(
    Map<String, dynamic> responseData,
    AuthUser user,
  ) async {
    final registration =
        responseData['deviceRegistration'] as Map<String, dynamic>? ?? const {};
    final offlineToken = registration['offlineToken'] as String?;

    if (offlineToken == null || offlineToken.isEmpty) {
      await _auth.clearTrustedDeviceAccessCache();
      return;
    }

    await _auth.replaceTrustedDeviceCredentials(
      offlineToken: offlineToken,
      user: user,
    );
  }

  Future<void> _seedOnboardingFinanceDefaults(
    OnboardingDraft draft,
    AuthUser user,
  ) async {
    final tenantId = user.tenantId;
    final schoolId = user.schoolId;
    if (tenantId == null || schoolId == null || draft.feeCategories.isEmpty) {
      return;
    }

    final scope = LocalDataScope(
      tenantId: tenantId,
      schoolId: schoolId,
      campusId: user.campusId,
    );
    final existingCategories = await _db.getFeeCategories(scope: scope);
    if (existingCategories.isNotEmpty) {
      return;
    }

    final now = DateTime.now();
    await _db.transaction(() async {
      for (final category in draft.feeCategories) {
        final categoryId = _uuid.v4();
        await _db.upsertFeeCategory(
          FeeCategoriesCompanion(
            id: Value(categoryId),
            tenantId: Value(tenantId),
            schoolId: Value(schoolId),
            name: Value(category.name),
            billingTerm: Value(category.billingTerm),
            isActive: const Value(true),
            serverRevision: const Value(0),
            deleted: const Value(false),
            createdAt: Value(now),
            updatedAt: Value(now),
          ),
        );
        await _db.enqueueSyncChange(
          entityType: 'fee_category',
          entityId: categoryId,
          operation: 'create',
          payload: {
            'id': categoryId,
            'tenantId': tenantId,
            'schoolId': schoolId,
            'name': category.name,
            'billingTerm': category.billingTerm,
            'isActive': true,
          },
        );

        final feeStructureItemId = _uuid.v4();
        await _db.upsertFeeStructureItem(
          FeeStructureItemsCompanion(
            id: Value(feeStructureItemId),
            tenantId: Value(tenantId),
            schoolId: Value(schoolId),
            feeCategoryId: Value(categoryId),
            classLevelId: const Value(null),
            termId: const Value(null),
            amount: Value(category.defaultAmount),
            notes: const Value('Seeded from onboarding wizard'),
            serverRevision: const Value(0),
            deleted: const Value(false),
            createdAt: Value(now),
            updatedAt: Value(now),
          ),
        );
        await _db.enqueueSyncChange(
          entityType: 'fee_structure_item',
          entityId: feeStructureItemId,
          operation: 'create',
          payload: {
            'id': feeStructureItemId,
            'tenantId': tenantId,
            'schoolId': schoolId,
            'feeCategoryId': categoryId,
            'classLevelId': null,
            'termId': null,
            'amount': category.defaultAmount,
            'notes': 'Seeded from onboarding wizard',
          },
        );
      }
    });
  }
}
